### ISSUE TITLE
JoltGeneric6DOFJoint3D angular limit gizmo not correct

### ISSUE THREAD

**Issue by tlobig** [godot-jolt/godot-jolt#855]
while playing around with this addon, which I really like, I figured there is a small issue with the Jolt-version of the 6DOF joints.

The short version: The angular limit gizmos only show the actual limits correctly if the joint itself is rotated by 90 degrees.
![screenshot in editor](https://github.com/godot-jolt/godot-jolt/assets/10432347/551a1756-446d-4911-bf79-fc77e3ba8333)

![screenshot when running the scene, notice middle joint is not limiting angle as previewed with Gizmo](https://github.com/godot-jolt/godot-jolt/assets/10432347/fbf58aa4-3e14-47b7-a429-ee50fff95806)

steps to reproduce:
should be obvious, have a static body  and a rigid body connected by a Jolt 6DOF joint, set an angular limit that's relatively narrow to see the effect AND the gizmo, add a camera and run the scene  to see the angular limit is not solved as previewed by the gizmo.

minimal project to reproduce:
[jolt_gizmo_minimal.zip](https://github.com/godot-jolt/godot-jolt/files/15316517/jolt_gizmo_minimal.zip)

Godot-Jolt version: 0.12.
Godot version: 4.2.1 stable

**Comment by mihe** [comment]
Thank you for reporting this. This has been in the back of my mind for a while now, but I never got around to it.

I've spent a silly amount of time trying to make sense of these joint gizmos, ever since I first implemented them, which I realize in hindsight was largely a result of me getting tripped up by the way that single-body joints work in Godot Physics, which I for some reason felt the need to accommodate for (over two-body joints). 

With this behavior having changed in #725 I should rightfully have changed the gizmo accordingly, but forgot to.

Anyway, this should be fixed when #862 is merged. It now seems to line up nicely with Jolt's own debug visualization, as well as Nvidia Omniverse, which happens to use a very similar visualization in its USD Composer editor.

**Comment by tlobig** [comment]
extremely cool and swift execution. nicely done.

**Comment by tlobig** [comment]
just compiled the master - is this the intended display of the gizmos?
![grafik](https://github.com/godot-jolt/godot-jolt/assets/10432347/21ddedf0-9b1f-4be0-b0fb-bbe212dffa32)
![grafik](https://github.com/godot-jolt/godot-jolt/assets/10432347/1ee0c664-ab3b-42e9-8500-e58d7df8dc74)

i.e. intuitively I would expect the gizmo to show the range of limits in the direction of the effect, but now it looks like its showing it on the opposite side.

**Comment by tlobig** [comment]
might be an issue with negative values - I just checked and it looks good with limits above 0°

**Comment by mihe** [comment]
Would you mind sharing the scene? It's hard to tell from the screenshots how you've set it up. It seemed to make sense in the scene I tried it with.

**Comment by tlobig** [comment]
you get the same issue with the minimal project up above. The underlying problem seems to be with where 0° starts? I phrased this as having to rotate the joint by 90° but then you fix only one axis and it gets messy when you need more than one axis limited.As you just changed the sign in your calculation i don't think that fixed it. I will see if I can suggest a fix on my own over then weekend.Am 23.05.2024 21:36 schrieb Mikael Hermansson ***@***.***>:
Would you mind sharing the scene/project? It's hard to tell from the screenshots how you've set it up. It seemed to make sense in the scene I tried it with.

**Comment by tlobig** [comment]
I'm pretty sure all we need is to add or substract half PI from the gizmo angle when calculating the points for it. However while trying this I don't see any effect on the gizmo drawing at all. I must be missing something crucial about the whole plugin build.

**Comment by tlobig** [comment]
sorry, bit pressed for time so I'll be sloppy.
basically you need to revert your change from #862, it's doing the wrong thing
instead line 83 in the jolt joint gizmo .cpp should be something like `const float angle = p_limit_lower + angle_step * (float)p_index  - Mathf_PI / 2.0;`
I also now noticed that the limits must be in range -180° to 180°, which propably follows naturally. It's just the UI doesn't limit this and maybe should.

**Comment by tlobig** [comment]
see suggested fix in #872

**Comment by mihe** [comment]
Yes, sorry, I saw it. I've been somewhat strapped for time when it comes to this extension in general lately, and this ended up pretty far down on my priority list when you in the PR mentioned that it wasn't a complete solution.

Frankly I'm become more tempted to just copy-paste Godot's own (equally strange) gizmo and call it a day. The fact that I deviated to begin with was somewhat questionable, given that the substitute joints are largely meant to just augment the functionality of the built-in ones. I figure if someone wants it to be different it should rightfully be fixed in Godot itself as well in any case.

Thank you for the PR though. I'll see what I end up doing about this. I'll go ahead and close your PR for now though.

**Comment by mihe** [comment]
With this extension now having been merged into Godot in the form of an engine module (see [README](https://github.com/godot-jolt/godot-jolt?tab=MIT-1-ov-file)), this extension will be considered to be in maintenance mode going forward, where only bug fixes will be considered for merging.

While this issue could still be considered a bug, it's highly unlikely that I'll spend any time addressing this, so I'll go ahead and close this as `wontfix`.
